<a href="https://colab.research.google.com/github/uttoxi/IBM-AI-Specialization/blob/main/Part_A_Building_neural_networks_with_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Peer-graded Assignment: Build a Regression Model in Keras (Part A)**


##In this part we use the dataset "Concrete_data.csv" and build a Regression model using Keras module

##**Stage-1 : Data processing and cleaning**

### Step-1:Importing pandas and numpy libraries

In [ ]:
import pandas as pd
import numpy as np

###Step-2 : Loading Data and generate descriptive statistics


The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them.

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate



In [ ]:
concrete_data = pd.read_csv('/content/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

**Interpretation of Datapoint:**
Concrete sample comprising of  **540 cubic meter** of cement,

**0 cubic meter** of blast furnace slag,

**0 cubic meter** of fly ash,

**162 cubic meter** of water,

**2.5 cubic meter** of superplaticizer,

**1040 cubic meter** of coarse aggregate,

**676 cubic meter** of fine aggregated, aged 28 days old,

has a **compressive strength of 79.99 MPa**

In [ ]:
concrete_data.shape

(1030, 9)

Generating descriptive statistics

In [ ]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

### Step-3 : Data cleaning- checking for nulls

In [ ]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

###Step-4 : Copying Independent variables(columns) in predictors dataframe and dependent or target variable "strength" in target variable

We are required to predict the "Strength" so "Strength is our Target and all other columns are predictors

In [ ]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength'] # Strength column

In [ ]:
predictors.head()


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [ ]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

In [ ]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

##**Stage-2: Building Keras model**

###Step-1 : Importing keras library and importing Sequential model and Dense layers

In [ ]:
import keras

**Notes:**

The **Sequential model** is used to create a linear stack of layers, making it easy to build and configure a neural network by adding layers one by one.

The **Dense layer** represents a fully connected (or densely connected) layer in the neural network, where each neuron in the layer is connected to every neuron in the previous and next layers. It's a fundamental building block for constructing feedforward neural networks.

In [ ]:
from keras.models import Sequential
from keras.layers import Dense

###Step-2: Defining the regression model with "**adam**" as optimizer and "**Relu**" Activation function and "**MSE**" as loss function

Following function creates a **Sequential** model that has **1 hidden layer** with **10 neurons** and a **ReLU activation function**. It uses the **adam optimizer** and the **mean squared error as the loss function**

In [ ]:
# define regression model
def regression_model():

    model = Sequential() # Creting Sequential model instance
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))


    model.compile(optimizer='adam', loss='mean_squared_error')  # compile model
    return model

###Step-3 : Importing the train_test_split from sklearn to randomely divide dataset in traing and testing dataset in 70:30 ratio

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)


##**Stage-3: Training and Testing the network**

###Step-1 : CCalling function "Regression_model" to create model

In [ ]:
# Creating instance of regression model
model = regression_model()

###Step-2 :Traing the model with 50 epochs

In [ ]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)


Epoch 1/50
23/23 - 0s - loss: 307757.3125 - 475ms/epoch - 21ms/step
Epoch 2/50
23/23 - 0s - loss: 232535.9219 - 25ms/epoch - 1ms/step
Epoch 3/50
23/23 - 0s - loss: 172916.6406 - 34ms/epoch - 1ms/step
Epoch 4/50
23/23 - 0s - loss: 126252.7031 - 30ms/epoch - 1ms/step
Epoch 5/50
23/23 - 0s - loss: 89619.3203 - 26ms/epoch - 1ms/step
Epoch 6/50
23/23 - 0s - loss: 61299.9453 - 25ms/epoch - 1ms/step
Epoch 7/50
23/23 - 0s - loss: 40185.1250 - 27ms/epoch - 1ms/step
Epoch 8/50
23/23 - 0s - loss: 25146.2715 - 27ms/epoch - 1ms/step
Epoch 9/50
23/23 - 0s - loss: 15096.5273 - 27ms/epoch - 1ms/step
Epoch 10/50
23/23 - 0s - loss: 8788.7891 - 26ms/epoch - 1ms/step
Epoch 11/50
23/23 - 0s - loss: 5256.1260 - 30ms/epoch - 1ms/step
Epoch 12/50
23/23 - 0s - loss: 3363.1021 - 28ms/epoch - 1ms/step
Epoch 13/50
23/23 - 0s - loss: 2460.4417 - 41ms/epoch - 2ms/step
Epoch 14/50
23/23 - 0s - loss: 2056.6792 - 25ms/epoch - 1ms/step
Epoch 15/50
23/23 - 0s - loss: 1870.6488 - 28ms/epoch - 1ms/step
Epoch 16/50
23/23 -

###Step-3 : Model evaluation on test data for losses

In [ ]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step


544.423095703125

###Step-4 : Calculating Mean Squared Error between Predicted and Actual Concrete strength

In [ ]:
from sklearn.metrics import mean_squared_error


In [ ]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

544.4231302139614 0.0


### Step-5: Repeating previous steps 50 times, i.e., create a list of 50 mean squared errors, and reporting

In [ ]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)



MSE 1: 170.9258270263672
10/10 [==============================] - 0s 1ms/step
MSE 2: 123.44837188720703
10/10 [==============================] - 0s 1ms/step
MSE 3: 111.46601867675781
10/10 [==============================] - 0s 1ms/step
MSE 4: 121.23660278320312
10/10 [==============================] - 0s 2ms/step
MSE 5: 121.6060791015625
10/10 [==============================] - 0s 2ms/step
MSE 6: 108.83181762695312
10/10 [==============================] - 0s 1ms/step
MSE 7: 132.5647735595703
10/10 [==============================] - 0s 1ms/step
MSE 8: 98.79930114746094
10/10 [==============================] - 0s 1ms/step
MSE 9: 129.0413055419922
10/10 [==============================] - 0s 1ms/step
MSE 10: 109.4228515625
10/10 [==============================] - 0s 1ms/step
MSE 11: 102.53578186035156
10/10 [==============================] - 0s 1ms/step
MSE 12: 100.44866180419922
10/10 [==============================] - 0s 1ms/step
MSE 13: 116.18551635742188
10/10 [========================

In [ ]:
print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))



Below is the mean and standard deviation of 50 mean squared errors with normalized data. Total number of epochs for each training is: 50

Mean: 114.00107623913013
Standard Deviation: 11.938904442986567


Part-A

Mean of MSE :  **114.00**

Standard Deviation of MSE : **11.94**